R/network.R

```diff
@@ -0,0 +1,301 @@
+}
+
+insert_intersection <- function(edge_pts, point, line_id) {
+  line_pts <- subset(edge_pts, edge_pts$linestring_id == line_id)
+  pt_x <- point[[1]]
+  pt_y <- point[[2]]
+  is_point_in_line <- nrow(
+    subset(line_pts, line_pts$x == pt_x & line_pts$y == pt_y)
+  ) >= 1
+  if (!is_point_in_line) {
+    startpoint <- subset(line_pts, line_pts$is_startpoint == TRUE)
+    kk <- as.numeric(rownames(startpoint))
+    w_break <- FALSE
+    while (!w_break) {
+      # Consider the line segments a - b.
+      # x is a valid intersection if the following condition is true:
+      # distance(a, b) == distance(a, x) + distance(x, b)  # nolint
+      pt_a_x <- edge_pts[kk, ]$x
+      pt_a_y <- edge_pts[kk, ]$y
+      pt_b_x <- edge_pts[kk + 1, ]$x
+      pt_b_y <- edge_pts[kk + 1, ]$y
+      d_ab <- distance(pt_a_x, pt_a_y, pt_b_x, pt_b_y)
+      d_ax <- distance(pt_a_x, pt_a_y, pt_x, pt_y)
+      d_bx <- distance(pt_b_x, pt_b_y, pt_x, pt_y)
+      is_intersection <- dplyr::near(d_ab, d_ax + d_bx, tol = 1.e-3)
+      if (is_intersection) {
+        insertion <- tibble::tibble_row(
+          sfg_id = line_id,
+          linestring_id = line_id,
+          x = pt_x,
+          y = pt_y,
+          is_startpoint = FALSE,
+          is_endpoint = FALSE
+        )
+        edge_pts <- tibble::add_row(edge_pts, insertion, .after = kk)
+        w_break <- TRUE
+      } else {
+        if (edge_pts[kk + 1, ]$is_endpoint) {
+          warning("point is not added to the edge df.")
+          w_break <- TRUE
+        }
+      }
+      kk <- kk + 1
+    }
+  }
+  edge_pts
+}
+
+#' Clean a spatial network.
+#'
+# nolint start
+#' Subdivide edges by [adding missing nodes](https://luukvdmeer.github.io/sfnetworks/articles/sfn02_preprocess_clean.html#subdivide-edges),
+#' remove [pseudo-nodes](https://luukvdmeer.github.io/sfnetworks/articles/sfn02_preprocess_clean.html#smooth-pseudo-nodes),
+#' simplify the network (see [simplify_network()]), and discard all but the
+#' main connected component.
+# nolint end
+#'
+#' @param network A network object
+#'
+#' @return A cleaned network object
+#' @export
+clean_network <- function(network) {
+  network |>
+    # subdivide edges by adding missing nodes
+    tidygraph::convert(sfnetworks::to_spatial_subdivision, .clean = TRUE) |>
+    # remove pseudo-nodes
+    tidygraph::convert(sfnetworks::to_spatial_smooth, .clean = TRUE) |>
+    # run simplification steps
+    simplify_network() |>
+    # keep only the main connected component of the network
+    tidygraph::activate("nodes") |>
+    dplyr::filter(tidygraph::group_components() == 1)
+}
+
+#' Simplify a spatial network by removing multiple edges and loops.
+#'
+# nolint start
+#' Simplify the graph, removing loops and double-edge connections following
+#' [this approach](https://luukvdmeer.github.io/sfnetworks/articles/sfn02_preprocess_clean.html#simplify-network).
+#' When dropping multiple edges, keep the shortest ones.
+# nolint end
```

Comment:
```suggestion
#' Simplify the graph, removing loops and double-edge connections following
#' \href{https://luukvdmeer.github.io/sfnetworks/articles/\
#' sfn02_preprocess_clean.html#simplify-network}{this approach}.
#' When dropping multiple edges, keep the shortest ones.
```

Reply:
Better approach, thanks! I will keep the markdown link however, it looks more readable to me..